server/aesdsocket.c

```diff
@@ -440,338 +472,347 @@ int main(int argc, char *argv[])
+		exit(EXIT_FAILURE);
+	}
 
-				/* Using memcpy */
-				//memcpy(storage_array+memcpy_counter, recv_buffer, i);
-				//memcpy_counter += packet_size;
-				//storage_array[packet_size] = '\n';
-				
-				/* Using strncpy */
-				strncat(storage_array, recv_buffer, i);
+	int fd = open(filepath, O_CREAT|O_RDWR|O_APPEND, 0644);
+	//int fd = open(filepath, O_WRONLY | O_APPEND);
+	if(fd == -1)
+	{
+		syslog(LOG_ERR, "ERROR: open() : %s \n", strerror(errno));
+		//syslog(LOG_ERR, "ERROR: open()\n");
+		free(storage_array);
+		exit(EXIT_FAILURE);
+	}
 
-				// Set the receive buffer to 0
-				memset(recv_buffer, 0, RECV_SIZE);
-			}
+	//pthread_mutex_lock(params->mutex);
+	int bytes_wriiten = write(fd, storage_array, packet_size);
+	if( bytes_wriiten == -1)
+	{
+		syslog(LOG_ERR, "ERROR: write() : %s \n", strerror(errno));
+		//syslog(LOG_ERR, "ERROR: read()\n");
+		
+	}
 
-		}
-		// Update total_data_size
-		total_data_size += packet_size;
-
-		/* For debugging */
-		// printf("Packet Size =  %d\n", packet_size);
-		// printf("total_data_size =  %d\n", total_data_size);
-		// printf("Storage array = \n%s", storage_array);
-		int status;
-		/* Write to file */
-		status = pthread_mutex_lock(&mutex_lock);
-		if(status)
+	//fdatasync(fd);
+	
+	//pthread_mutex_unlock(params->mutex);
+	//saclose(fd);
+	// Open the file to read
+	//fd = open(filepath, O_RDONLY);
+	// if(fd < 0)
+	// {
+	// 	syslog(LOG_ERR, "ERROR: open()\n");
+	// 	free(storage_array);
+	// 	exit(EXIT_FAILURE);
+	// }
+	lseek(fd, 0, SEEK_SET);
+	
+	/* Logic to read one byte from the file and send one byte at a time to client*/
+	
+	//int j = 0;
+	char a;
+	//while(j < total_data_size)
+	//pthread_mutex_lock(params->mutex);
+	// reading doesnt need a lock
+	while((nread = read(fd, &a, 1)) != 0)
+	{
+		//nread = read(fd, &a, 1);
+		if( nread == -1)
 		{
-			syslog(LOG_ERR, "ERROR: mutex_lock() fail");
+			syslog(LOG_ERR, "ERROR: read() : %s \n", strerror(errno));
+			free(storage_array);
 			exit(EXIT_FAILURE);
 		}
 
-		int fd = open(filepath, O_CREAT|O_RDWR|O_APPEND, 0644);
-		//int fd = open(filepath, O_WRONLY | O_APPEND);
-		if(fd < 0)
+		status = send(params->client_sock_fd, &a, 1, 0);
+		if(status == -1)
 		{
-			syslog(LOG_ERR, "ERROR: open()\n");
-			free(storage_array);
+			syslog(LOG_ERR, "ERROR: send() : %s \n", strerror(errno));
+			//syslog(LOG_ERR, "ERROR: mutex_lock() fail");
 			exit(EXIT_FAILURE);
 		}
-		//pthread_mutex_lock(params->mutex);
-		int bytes_wriiten = write(fd, storage_array, packet_size);
-		if( bytes_wriiten < 0)
-		{
-			syslog(LOG_ERR, "ERROR: read()\n");
-		}
+		//j++;
+	}
 
-		//fdatasync(fd);
-		
-		//pthread_mutex_unlock(params->mutex);
-		//saclose(fd);
-		// Open the file to read
-		//fd = open(filepath, O_RDONLY);
-		// if(fd < 0)
-		// {
-		// 	syslog(LOG_ERR, "ERROR: open()\n");
-		// 	free(storage_array);
-		// 	exit(EXIT_FAILURE);
-		// }
-		lseek(fd, 0, SEEK_SET);
+	close(fd);
+	status = pthread_mutex_unlock(&mutex_lock);
```

Comment:
@sanishkharade  Are you not holding mutex lock for too long? this is locking code and data